flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/OperatorEventDispatcherImpl.java

```diff
@@ -0,0 +1,124 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.streaming.runtime.tasks;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.runtime.execution.Environment;
+import org.apache.flink.runtime.jobgraph.OperatorID;
+import org.apache.flink.runtime.jobgraph.tasks.OperatorEventSender;
+import org.apache.flink.runtime.operators.coordination.OperatorEvent;
+import org.apache.flink.runtime.operators.coordination.OperatorEventDispatcher;
+import org.apache.flink.runtime.operators.coordination.OperatorEventGateway;
+import org.apache.flink.runtime.operators.coordination.OperatorEventHandler;
+import org.apache.flink.util.FlinkException;
+import org.apache.flink.util.FlinkRuntimeException;
+import org.apache.flink.util.SerializedValue;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/**
+ * An implementation of the {@link OperatorEventDispatcher}.
+ *
+ * <p>This class is intended for single threaded use from the stream task mailbox.
+ */
+@Internal
+final class OperatorEventDispatcherImpl implements OperatorEventDispatcher {
+
+	private final Map<OperatorID, OperatorEventHandler> handlers;
+
+	private final ClassLoader classLoader;
+
+	private final OperatorEventSender toCoordinator;
+
+	OperatorEventDispatcherImpl(ClassLoader classLoader, OperatorEventSender toCoordinator) {
+		this.classLoader = checkNotNull(classLoader);
+		this.toCoordinator = checkNotNull(toCoordinator);
+		this.handlers = new HashMap<>();
+	}
+
+	void dispatchEventToHandlers(OperatorID operatorID, SerializedValue<OperatorEvent> serializedEvent) throws FlinkException {
+		final OperatorEvent evt;
+		try {
+			evt = serializedEvent.deserializeValue(classLoader);
+		}
+		catch (IOException | ClassNotFoundException e) {
+			throw new FlinkException("Could not deserialize operator event", e);
+		}
+
+		final OperatorEventHandler handler = handlers.get(operatorID);
+		if (handler != null) {
+			handler.handleOperatorEvent(evt);
+		}
+		else {
+			throw new FlinkException("Operator not registered for operator events");
+		}
+	}
+
+	@Override
+	public OperatorEventGateway registerEventHandler(OperatorID operator, OperatorEventHandler handler) {
+		final OperatorEventGateway gateway = new OperatorEventGatewayImpl(toCoordinator, operator);
+		final OperatorEventHandler prior = handlers.putIfAbsent(operator, handler);
+
+		if (prior == null) {
+			return gateway;
+		}
+		else {
+			throw new IllegalStateException("already a handler registered for this operatorId");
+		}
+	}
+
+	// ------------------------------------------------------------------------
+
+	static OperatorEventDispatcherImpl fromEnvironment(Environment env) {
```

Comment:
It seems we only use this method to construct an `OperatorEventDispatcherImpl`, should we just use have a constructor taking the `Environment` and remove this method?

Reply:
I think we suggest in the coding guidelines to avoid "constructor doing the work", as it usually leads to less maintainable designs. For a constructor that does work, a factory method is the clean way to go, in my opinion.

Reply:
I am not sure. The static method here seems a little verbose to me. If we don't want the constructor to call getter methods on the arguments, maybe the caller can simply call the constructor directly.

Arguably, this method makes the code slightly more readable, but in that case, it seems better to define it as a private method in the `OperatorChain`.

Reply:
Fair enough. I have no strong feelings about this particular issue, so if you believe it is better to remove the static method, let me do that.